webcam-capture-drivers/driver-ffmpeg-cli/src/main/java/com/github/sarxos/webcam/ds/ffmpegcli/FFmpegCliDevice.java

```diff
@@ -1,260 +1,218 @@
-		} catch (FileNotFoundException e) {
-			throw new RuntimeException(e);
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		} catch (InterruptedException e) {
-			throw new RuntimeException(e);
-		}
-	}
-
-	@Override
-	public synchronized void close() {
-
-		if (!open.compareAndSet(true, false)) {
-			return;
-		}
-
-		try {
-			dis.close();
-		} catch (IOException e) {
-			throw new RuntimeException(e);
-		}
-
-		process.destroy();
-
-		try {
-			process.waitFor();
-		} catch (InterruptedException e) {
-			throw new RuntimeException(e);
-		}
-
-		if (!pipe.delete()) {
-			pipe.deleteOnExit();
-		}
-	}
-
-	@Override
-	public void dispose() {
-		if (disposed.compareAndSet(false, true) && open.get()) {
-			close();
-		}
-	}
-
-	@Override
-	public boolean isOpen() {
-		return open.get();
-	}
-
-	@Override
-	public String toString() {
-		return "video device " + name;
-	}
+    private static final Logger LOG = LoggerFactory.getLogger(FFmpegCliDevice.class);
+
+    private Process process = null;
+
+    private File vfile = null;
+    private String name = null;
+    private Dimension[] resolutions = null;
+    private Dimension resolution = null;
+
+    private AtomicBoolean open = new AtomicBoolean(false);
+    private AtomicBoolean disposed = new AtomicBoolean(false);
+
+    BufferedImage image;
+
+    protected FFmpegCliDevice(File vfile, String vinfo) {
+        String[] parts = vinfo.split(" : ");
+
+        this.vfile = vfile;
+        this.name = vfile.getAbsolutePath();
+        this.resolutions = readResolutions(parts[3].trim());
+
+        Dimension resolution = getResolution();
+        image = new BufferedImage(resolution.width, resolution.height, BufferedImage.TYPE_3BYTE_BGR);
+    }
+
+    public void startProcess() throws IOException {
+        ProcessBuilder builder = new ProcessBuilder(buildCommand());
+        builder.redirectErrorStream(true); // so we can ignore the error stream
+
+        Process process = builder.start();
+        InputStream out = process.getInputStream();
+        OutputStream in = process.getOutputStream();
+
+        int SIZE = resolution.width * resolution.height * 3;
+        int CHUNK_SIZE = SIZE / 20;
+
+        int cursor = 0;
+
+        byte[] chunk = new byte[CHUNK_SIZE];
+        ByteBuffer buffer = ByteBuffer.allocate(SIZE);
+
+        while (isAlive(process)) {
+            int no = out.available();
+            if (no >= chunk.length) {
+                int n = out.read(chunk, 0, chunk.length);
+                cursor += n;
+
+                // If buffer is not full yet
+                if (cursor < SIZE) {
+                    // Append the chunk to the buffer
+                    buffer.put(chunk, 0, n);
+                } else {
+                    // Read Image
+                    byte[] bgr = buffer.array();
+                    image.setData(Raster.createRaster(image.getSampleModel(), new DataBufferByte(bgr, bgr.length), new Point()));
```

Comment:
Access to `image` should be synchronized. You are setting new data here and in `getImage()` you are returning it. There is no guarantee of data consistency between these operations. This should rather be:

```java
synchronized (image) {
  image.setData(Raster.createRaster(image.getSampleModel(), new DataBufferByte(bgr, bgr.length), new Point()));
}